### PR TITLE
Share field mapper tests with other modules

### PR DESCRIPTION
**Pull request by shans96**
WIP PR to close issue #92947

My approach for solving this problem was to see if I could get `UnsignedLongMapperTests` to extend `NumberFieldMapperTests` by copying `NumberFieldMapperTests` over to the test framework (and then do the same for `ScaledFloatFieldMapperTests`). Once I've made `UnsignedLongMapperTests` extend `NumberFieldMapperTests`, my plan was to make it extend `WholeNumberFieldMapperTests` instead, as that seems more appropriate (and specialised) for the unsigned long type.

I have managed to make `UnsignedLongMapperTests` extend  `NumberFieldMapperTests` but there are now some test failures for:
-`testCoerce`
-`testOutOfRangeValues`
-`testAllowMultipleValuesField`
-`testIgnoreMalformedWithObject`
-`testScriptableTypes`

It looks like some wider changes are needed to fix the tests (e.g., `NumberFieldMapper` has to be updated with some code to support unsigned longs), but I'm not sure of whether this is the right decision or not from an architectural perspective, so I wanted to get some early feedback to make sure I'm on the right track.

Any insight on the above would be appreciated!

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java [hunk 1]
```diff
@@ -336,12 +336,12 @@ protected <T> T compileOtherScript(Script script, ScriptContext<T> context) {
             }
 
             @Override
-            ScriptFactory emptyFieldScript() {
+            protected ScriptFactory emptyFieldScript() {
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java:339]: This class wouldn't bee needed in the "server" module after the move I think. Small changes in visibility are okay if needed, please comment them if you have questions.

**Review by cbuescher — COMMENTED** [review]
Hi @shans96, I took a quick look at this PR but didn't look into all the details because you still marked it as draft.
The general idea is right, I left some comments on some specifics.
I think the general ask with this issue is to move the testing code and put the two mentioned tests (UnsignedLongFieldMapperTests, ScaledFloatFieldMapperTests) under one of the more specific base tests. I think that when tests of the base class fail (e.g. "coerce" doesn't seem to be supported in UnsignedLongFieldMapper), we should overwrite or disable those tests in the inheriting test class. If only some assertions need to be fixed, that's okay of course, but I would like leave behavioural changes out of this refactoring PR as much as possible.
Hope that helps, let me know if you need more input.

File: modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapperTests.java [hunk 2]
```diff
@@ -196,7 +197,7 @@ public void testStore() throws Exception {
         assertEquals(1230, storedField.numericValue().longValue());
     }
 
-    public void testCoerce() throws Exception {
+    public void testCoerce() throws IOException {
```
Comment (shans96) [inline modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapperTests.java:200]: Changed to make the method signature match how it is declared in `NumberFieldMapperTests`

File: test/framework/src/main/java/org/elasticsearch/index/mapper/IngestScriptSupport.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+import org.elasticsearch.script.Script;
+import org.elasticsearch.script.ScriptContext;
+import org.elasticsearch.script.ScriptFactory;
+
+import java.util.Optional;
+
+// Copied from MapperTestCase
+// TODO: Is it fine to keep this class in the index package and keep the weaker access levels for it, or should it be moved somewhere else?
+public abstract class IngestScriptSupport {
```
[the remark this answers is not in the document]
Comment (shans96) [reply test/framework/src/main/java/org/elasticsearch/index/mapper/IngestScriptSupport.java]: Good point, thanks- don't think it was needed anymore

File: test/framework/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java [hunk 4]
```diff
@@ -315,7 +317,7 @@ public void testMetricAndDocvalues() {
     }
 
     @Override
-    protected final Object generateRandomInputValue(MappedFieldType ft) {
+    protected Object generateRandomInputValue(MappedFieldType ft) {
```
Comment (shans96) [inline test/framework/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java:320]: Needs to be made non-final so `UnsignedLongFieldMapperTests` and `ScaledFloatFieldMapperTests` can override it

File: modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapperTests.java [hunk 5]
```diff
@@ -432,6 +433,39 @@ protected IngestScriptSupport ingestScriptSupport() {
         throw new AssumptionViolatedException("not supported");
     }
 
+    @Override
+    protected List<NumberFieldMapperTests.OutOfRangeSpec> outOfRangeSpecs() {
+        return Collections.emptyList(); // unimplemented
+    }
+
+    @Override
+    public void testIgnoreMalformedWithObject() {} // unimplemented
```
Comment (shans96) [inline modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapperTests.java:442]: I've overridden the methods for now but I'm sure there's a cleaner way to write these. I've seen that some tests can be skipped in gradle, would that be cleaner, and if so, where should I look to find an example?

File: test/framework/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java [hunk 6]
```diff
@@ -459,4 +461,30 @@ public List<SyntheticSourceInvalidExample> invalidExample() throws IOException {
             );
         }
     }
+
+    // Copied from NumberFieldTypeTests, better to move this into a class of its own?
+    public static class OutOfRangeSpec {
```
Comment (shans96) [inline test/framework/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java:466]: Feedback appreciated on this one

File: x-pack/plugin/mapper-unsigned-long/src/test/java/org/elasticsearch/xpack/unsignedlong/UnsignedLongFieldMapperTests.java [hunk 7]
```diff
@@ -365,6 +370,38 @@ protected IngestScriptSupport ingestScriptSupport() {
...
+        if (randomBoolean()) {
+            return randomLong();
+        }
+        if (randomBoolean()) {
+            return randomDouble();
+        }
+        assumeFalse("https://github.com/elastic/elasticsearch/issues/70585", true);
+        // TODO: unsigned long has a max value of (2^64)-1, but it can't be used in the randomDoubleBetween method because it's too large
+        return randomDoubleBetween(0L, Long.MAX_VALUE, true);
+    }
```
Comment (shans96) [inline x-pack/plugin/mapper-unsigned-long/src/test/java/org/elasticsearch/xpack/unsignedlong/UnsignedLongFieldMapperTests.java:403]: Still need to resolve this one as well, was thinking of making a new method that supports larger numbers but maybe there's an easier way to support this for unsigned longs?

**Comment by shans96** [comment]
Hi @cbuescher, thanks for the feedback. I've made changes based on your comments and think the PR is ready for a proper review; `UnsignedLongFieldMapperTests` now extends `WholeNumberFieldMapperTests` and `ScaledFloatFieldMapperTests` now extends `NumberFieldMapperTests`. I had some small questions on making the code cleaner where I've overridden the methods, and some other small questions, but hopefully they are quick to resolve :+1: 

Look forward to hearing from you!

**Review by cbuescher — CHANGES_REQUESTED** [review]
Sorry for the late reply. I left a few comments but nothing big, I think we should merge this and work on extending the tests to fill in the missing methods later. As it is, this already improves things by using the same base test class and doesn't remove any coverage. Can you update the branch and add the few things I suggested before I run our CI test suite please?

**Comment by cbuescher** [comment]
@shans96 just checking in to see if you want to continue this PR, since the branch has gone stale a bit and would need updating. As I mentioned earlier in the comments I think there are only a few smaller changes left to address before we can merge this.

**Comment by shans96** [comment]
Hi @cbuescher, apologies for the delay in reply on this one. Keen to finish this PR and will make it a focus- I've updated the comments based on your feedback. I'm currently looking at making `OutOfRangeSpec` publicly accessible: adding a dependency on `elasticsearch.server.test` wouldn't make sense (and would cause a circular dependency I think). Thinking of moving the class to the test framework instead- would this make sense?

**Comment by cbuescher** [comment]
> Thinking of moving the class to the test framework instead- would this make sense

It think so, yes. Maybe also make the name a bit more specific then, since that class seems tied to the number type tests, so maybe "NumberTypeOutOfRangeSpec" or something along those lines would be good. Please also try to update the branch, I can help with that if needed.

**Comment by cbuescher** [comment]
Hi @shans96, I'll hand this over to @pmpailis while I'm out for a few weeks. I we only need updating the branch and move the shared "OutOfRangeSpec" code somewhere where it can be accessed by modules. That is, if all tests pass in CI.

**Comment by shans96** [comment]
Hi, apologies again for the delay in updates on this one- it's been a hectic holiday season. I've moved `OutOfRangeSpec` into the test framework under the name `NumberTypeOutOfRangeSpec`- not sure if the class location/name are ideal, so any feedback on this one would be appreciated. I've also updated this branch with the latest changes from `master` and everything builds successfully on my local machine, so I think it's ready for CI tests.

**Comment by pmpailis** [comment]
Thanks for continuing this work @shans96 - changes LGTM! 
The only thing pending prior to running CI is to address the conflicts with `main` as some of the files above were changed recently (e.g. [here](https://github.com/elastic/elasticsearch/pull/103510/files#diff-1a110e1edc49686c15688a028b756a0a4d3cebffae59e5e3d85ae3e0ca76cf37R370)). Happy to help with addressing this if needed :)

**Comment by pmpailis** [comment]
@shans96 happy to assist with merging/resolving conflicts to move this forward :)

**Comment by shans96** [comment]
Hi, thanks for the ping :+1: I've just merged the latest version of `main` into my branch, but haven't run into any merge conflicts. I was anticipating some, so I looked at the PR you linked and tried to search for the new method declarations in the latest version of `main`, but they don't seem to exist there either. Wondering if I missed something, or were there additional changes afterwards that kept the field mapper tests as they were originally?

**Comment by pmpailis** [comment]
Thanks for taking a look @shans96 ! Yeah, actually it seems that some changes were reverted in [a later PR](https://github.com/elastic/elasticsearch/pull/104218/files#diff-031008a5bee59ebc302cdd947cd0259f5aa716cb49e4a69efd73cd968956b85dR385) a few days ago, so all seems fine now :)  Will proceed with running CI tests :)

**Comment by pmpailis** [comment]
@elasticmachine test this please

**Comment by pmpailis** [comment]
buildkite test this please

**Review by pmpailis — APPROVED** [review]

**Comment by pmpailis** [comment]
Thank you for your effort & contribution @shans96 ! CI tests have all passed 🚀  so I'll proceed with merging :)

**Pull request merged**